src/monitor/node_active_protocol.c

```diff
@@ -1172,34 +1332,112 @@ start_maintenance(PG_FUNCTION_ARGS)
 
 	if (IsCurrentState(currentNode, REPLICATION_STATE_PRIMARY))
 	{
+		List *standbyNodesGroupList = AutoFailoverOtherNodesList(currentNode);
+		AutoFailoverNode *firstStandbyNode = linitial(standbyNodesGroupList);
+		char message[BUFSIZE] = { 0 };
+
 		/*
-		 * We put the primary to prepare_maintenance now, and the secondary to
-		 * prepare_replication.
+		 * Set the primary to prepare_maintenance now, and if we have a single
+		 * secondary we assign it prepare_promotion, otherwise we need to elect
+		 * a secondary, same as in perform_failover.
 		 */
 		LogAndNotifyMessage(
 			message, BUFSIZE,
-			"Setting goal state of %s:%d to prepare_maintenance "
-			"and %s:%d to prepare_promotion "
+			"Setting goal state of node %d (%s:%d) to prepare_maintenance "
 			"after a user-initiated start_maintenance call.",
-			currentNode->nodeHost, currentNode->nodePort,
-			otherNode->nodeHost, otherNode->nodePort);
+			currentNode->nodeId, currentNode->nodeHost, currentNode->nodePort);
 
 		SetNodeGoalState(currentNode,
 						 REPLICATION_STATE_PREPARE_MAINTENANCE, message);
 
-		SetNodeGoalState(otherNode,
-						 REPLICATION_STATE_PREPARE_PROMOTION, message);
+		if (totalNodesCount == 2)
+		{
+			AutoFailoverNode *otherNode = firstStandbyNode;
+
+			/*
+			 * We put the only secondary node straight to prepare_replication.
+			 */
+			LogAndNotifyMessage(
+				message, BUFSIZE,
+				"Setting goal state of %s:%d to prepare_maintenance "
+				"and %s:%d to prepare_promotion "
+				"after a user-initiated start_maintenance call.",
+				currentNode->nodeHost, currentNode->nodePort,
+				otherNode->nodeHost, otherNode->nodePort);
+
+			SetNodeGoalState(otherNode,
+							 REPLICATION_STATE_PREPARE_PROMOTION, message);
+		}
+		else
+		{
+			/* now proceed with the failover, starting with the first standby */
+			(void) ProceedGroupState(firstStandbyNode);
+		}
+
+		PG_RETURN_BOOL(true);
+	}
+
+	/*
+	 * Done with dealing with a primary, we now have a secondary to take care
+	 * of. The primary needs to be in a stable state to allow for maintenance.
+	 */
+	primaryNode = GetPrimaryNodeInGroup(currentNode->formationId,
+										currentNode->groupId);
+
+	if (primaryNode == NULL)
+	{
+		ereport(ERROR,
+				(errmsg("couldn't find the primary node in formation \"%s\", "
+						"group %d",
+						currentNode->formationId, currentNode->groupId)));
+	}
+
+	/*
+	 * We need to always have at least formation->number_sync_standbys nodes in
+	 * the SECONDARY state, otherwise writes may be blocked on the primary. So
+	 * we refuse to put a node in maintenance when it would force blocking
+	 * writes.
+	 */
+	secondaryNodesList =
+		AutoFailoverOtherNodesListInState(primaryNode,
+										  REPLICATION_STATE_SECONDARY);
+
+	secondaryNodesCount = list_length(secondaryNodesList);
+
+	if (formation->number_sync_standbys > 0 &&
```

Comment:
I guess we need a similar check on `if (IsCurrentState(currentNode, REPLICATION_STATE_PRIMARY))` case as well. 


Say we have 1+3 cluster in the following configuration and sync_standbys = 1: 
```
show state --pgdata /Users/onderkalaci/Documents/pg_auto_failover/monitor/
  Name |         Host:Port |  Node |     Current State |    Assigned State | Priority | Quorum
-------+-------------------+-------+-------------------+-------------------+----------+-------
node_1 | 192.168.2.5:10001 |   0/1 |           primary |           primary |       50 |      t
node_2 | 192.168.2.5:10002 |   0/2 |       maintenance |       maintenance |       50 |      t
node_3 | 192.168.2.5:10003 |   0/3 |         secondary |         secondary |       50 |      t
node_4 | 192.168.2.5:10004 |   0/4 |       maintenance |       maintenance |       50 |      t
```

There seems nothing to prevent moving the primary to maintenance:
```
 pg_autoctl enable maintenance --pgdata /Users/onderkalaci/Documents/pg_auto_failover/worker1c --allow-failover
pg_autoctl enable maintenance --pgdata /Users/onderkalaci/Documents/pg_auto_failover/worker1a --allow-failover
15:12:08 32769 INFO  Listening monitor notifications about state changes in formation "default" and group 0
15:12:08 32769 INFO  Following table displays times when notifications are received
    Time |  ID |        Host |   Port |       Current State |      Assigned State
---------+-----+-------------+--------+---------------------+--------------------
15:12:09 |   1 | 192.168.2.5 |  10001 |             primary | prepare_maintenance
15:12:09 |   1 | 192.168.2.5 |  10001 | prepare_maintenance | prepare_maintenance
15:12:09 |   3 | 192.168.2.5 |  10003 |           secondary |          report_lsn
15:12:15 |   3 | 192.168.2.5 |  10003 |          report_lsn |          report_lsn
15:12:15 |   3 | 192.168.2.5 |  10003 |          report_lsn |   prepare_promotion
15:12:16 |   3 | 192.168.2.5 |  10003 |   prepare_promotion |   prepare_promotion
15:12:16 |   3 | 192.168.2.5 |  10003 |   prepare_promotion |    stop_replication
15:12:17 |   3 | 192.168.2.5 |  10003 |    stop_replication |    stop_replication
15:12:17 |   3 | 192.168.2.5 |  10003 |    stop_replication |        wait_primary
15:12:17 |   1 | 192.168.2.5 |  10001 | prepare_maintenance |         maintenance
15:12:17 |   3 | 192.168.2.5 |  10003 |        wait_primary |        wait_primary
15:12:18 |   1 | 192.168.2.5 |  10001 |         maintenance |         maintenance

 pg_autoctl show state --pgdata /Users/onderkalaci/Documents/pg_auto_failover/monitor/
  Name |         Host:Port |  Node |     Current State |    Assigned State | Priority | Quorum
-------+-------------------+-------+-------------------+-------------------+----------+-------
node_1 | 192.168.2.5:10001 |   0/1 |       maintenance |       maintenance |       50 |      t
node_2 | 192.168.2.5:10002 |   0/2 |       maintenance |       maintenance |       50 |      t
node_3 | 192.168.2.5:10003 |   0/3 |      wait_primary |      wait_primary |       50 |      t
node_4 | 192.168.2.5:10004 |   0/4 |       maintenance |       maintenance |       50 |      t

``` 
So, we're again in the not very optimal situation.

Reply:
I have now added a test case for this situation, and fixed the code to make it pass the test suite.